src/main/java/com/google/cloud/spanner/pgadapter/statements/BackendConnection.java

```diff
@@ -387,12 +397,41 @@ StatementResult executeOnSpanner(Statement statement) {
           && new SimpleParser(statement.getSql()).peekKeyword("insert")) {
         try {
           spannerConnection.setAutocommitDmlMode(AutocommitDmlMode.TRANSACTIONAL);
-          return spannerConnection.execute(statement);
+          return executeOnSpannerWithLogging(statement);
         } finally {
           spannerConnection.setAutocommitDmlMode(AutocommitDmlMode.PARTITIONED_NON_ATOMIC);
         }
       }
-      return spannerConnection.execute(statement);
+      return executeOnSpannerWithLogging(statement);
+    }
+
+    private StatementResult executeOnSpannerWithLogging(Statement statement) {
+      logger.log(
+          Level.FINER,
+          Logging.format(
+              "Executing",
+              Action.Starting,
+              () -> String.format("Statement: %s", statement.getSql())));
+      Stopwatch stopwatch = Stopwatch.createStarted();
+      StatementResult result = spannerConnection.execute(statement);
+      Duration executionDuration = stopwatch.elapsed();
+      logger.log(
+          Level.FINER,
+          Logging.format(
+              "Executing",
+              Action.Finished,
+              () -> String.format("Statement: %s", statement.getSql())));
+      if (executionDuration.compareTo(sessionState.getLogSlowStatementThreshold()) >= 0) {
```

Comment:
can this ever be null ?

Reply:
No, it has an implementation that will always return a default value.